Introduce the commands for API add, delete, update and generate deployment directory for Kubernetes (#3336)
wso2/docs-apim: en/docs/install-and-setup/setup/api-controller/managing-apis-in-kubernetes/managing-apis-in-kubernetes.md

# Managing APIs in Kubernetes

> WSO2 API Controller, **apictl** allows you to manage APIs in Kubernetes with the use of WSO2 API operator. This approach uses the kubernetes native way to deploy the APIS into a running microgateway cluster from the command line.

Reviewer: ```suggestion
**WSO2 API Controller (apictl)** allows you to manage APIs in Kubernetes with the use of WSO2 API Operator. This approach uses the Kubernetes native way to deploy the APIS into a running microgateway cluster from the command line.
```
Recently we decided to introduce **apictl** abbreviation as shown above when starting any apictl related doc. After that, you can use this abbreviation through the doc. We can do the same for API Operator as well if there is an abbreviation for that.

Author: Fixed with 21f466ce5d99225a24a45351eae0997e87f8c2f5

!!! Prerequisites

    - [Kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

    - [Kubernetes v1.14 or above](https://Kubernetes.io/docs/setup/) <br>

        - Minimum CPU : 2vCPU
        - Minimum Memory : 2GB

!!! info
    **Before you begin** 

>     -  First download the microgateway distribution from the [github release page](https://github.com/wso2/product-microgateway/releases/) and extract that to a folder of your choice. Then deploy microgateway in kubernetes cluster. Please refer [Steps for deploying  microgateway in kubernetes]()

Reviewer 2: Don't we need to add the docs link after this for the steps for the mentioned deployment?
**[Steps for deploying microgateway in kubernetes]**

Reviewer 2: And Shall we change the reference subject as  "Steps for deploying  Microgateway in Kubernetes" as @wasuradananjith mentioned.

Author: Fixed with 21f466ce5d99225a24a45351eae0997e87f8c2f5

>     -  Make sure the CTL is downloaded and initialized, if not follow the steps in [Download and Initialize the CTL Tool]({{base_path}}/install-and-setup/setup/api-controller/getting-started-with-wso2-api-controller/#download-and-initialize-the-ctl-tool).

Reviewer: ```suggestion
    -  Make sure the apictl is downloaded and initialized, if not follow the steps in [Download and initialize the apictl]({{base_path}}/install-and-setup/setup/api-controller/getting-started-with-wso2-api-controller/#download-and-initialize-the-apictl).
```
Please recheck the links. These have been changed recently.

Author: Fixed with 21f466ce5d99225a24a45351eae0997e87f8c2f5

>     -  Ensure that the API Operator is deployed in the kubernetes cluster. Please refer [API Operator documentation](https://github.com/wso2/k8s-api-operator/tree/master) for detailed steps.

Reviewer: ```suggestion
    -  Ensure that the API Operator is deployed in the Kubernetes cluster. Please refer [API Operator documentation](https://github.com/wso2/k8s-api-operator/tree/master) for detailed steps.
```

Reviewer: Let's make the first letter uppercase here **Kubernetes** ?

Author: Fixes with 21f466ce5d99225a24a45351eae0997e87f8c2f5

## Add an API to Kubernetes cluster

> 1.  Run the following command to add an API to the Kubernetes cluster using either a swagger file, zip file, or an API project folder. When you add the API to the Kubernetes cluster, the API operator successfully deploys the project zip to the WSO2 microgateway Adapter component.

Reviewer: ```suggestion
1.  Run the following command to add an API to the Kubernetes cluster using either a swagger file, zip file, or an API project folder. When you add the API to the Kubernetes cluster, the API operator successfully deploys the project zip to the WSO2 Microgateway Adapter component.
```
Let's make the first letter uppercase **Microgateway**?

Author: Fixed with 21f466ce5d99225a24a45351eae0997e87f8c2f5

        ```

## Generate Deployment Directory for Kubernetes

1.  Run the following command to generate a sample directory with all the contents to use as the deployment directory when performing CI/CD pipeline tasks in Kubernetes.

    -   **Command** 

        ```go
        apictl k8s gen deployment-dir -s <Path of Source directory> -d <Path of Destination directory>
        ```

        !!! info
            **Flags:**

            -   Required :  
                `--source` or `-s` : Path of the source directory to be used when generating the directory. Source can be a API project folder or a zipped API project.
            -   Optional :
                `--destination` or `-d` : Path of the destination directory where the directory should be generated.       
    
        !!! example
            ```go
             apictl k8s gen deployment-dir -s PizzaShackAPI-1.0.0
            ```
            ```go
             apictl k8s gen deployment-dir -s PizzaShackAPI-1.0.0 -d Deployment-Directory
            ```

    -   **Response**

        ```go
        The deployment directory for PizzaShackAPI-1.0.0 file is generated at Deployment-Directory
        ```

> 

Reviewer: Shall we mention what will be generated? The folder structure perhaps as we mentioned in the apictl - **Configuring Environment Specific Parameters** page.
![image](https://user-images.githubusercontent.com/25246848/111940878-d039c680-8af5-11eb-82c6-ae8906bb0575.png)

Author: Fixed with 3cbe17367ae5223e044356397e754ce547859314

2.  You can add [kustomization.yaml]({{base_path}}/install-and-setup/setup/api-controller/managing-apis-in-kubernetes/resources/kustomization.yaml), [kustomize-config.yaml]({{base_path}}/install-and-setup/setup/api-controller/managing-apis-in-kubernetes/resources/kustomize-config.yaml) and project zip file.

3.  Run the following command to generate the related config maps.

    -   **Command** 

        ```go
        kubectl kustomize

        ```

> 

Reviewer 2: Shall we remove the extra new line here?

Author: Fixed with 21f466ce5d99225a24a45351eae0997e87f8c2f5
